lib/internal/modules/esm/default_resolve.js

```diff
@@ -51,9 +51,10 @@ const extensionFormatMap = {
 };
 
 function resolve(specifier, parentURL) {
-  if (NativeModule.nonInternalExists(specifier)) {
+  const [maybeBuiltinSpecifier] = specifier.split(/\?|#/);
+  if (NativeModule.nonInternalExists(maybeBuiltinSpecifier)) {
     return {
-      url: specifier,
+      url: maybeBuiltinSpecifier,
       format: 'builtin'
     };
   }
```

Comment:
I don't think we should permit `?` and `#` for builtins, as these should behave similarly to any `import 'std:string'` style core modules in the browser in future as well, which very likely won't support hashes or querystrings.

Reply:
I'm fine either way, but would point out that if we treat builtins like CJS, there is an argument that we should allow those and redirect to the non-query/non-hash module record like other CJS modules.

Reply:
Actually the closer matching thing on the web is probably package maps. Note that we don't support for example `import 'lodash?query' for package lookups` and also package name maps wouldn't support this either in the browser. Package imports are NOT the same as URL imports.

Reply:
if we don't support it for builtin bare import, we can't support it for cjs bare import, and by extension esm bare import. there has to be consistency in some direction.

Reply:
@guybedford shouldn't the package name map still perform the request and then apply the request parameters to the response (similar to what happens to `import.meta.url` in browsers). You cannot map `lodash?query` but you should be able to import it I think from my understanding of package name maps.

Reply:
@guybedford is that intentional given that you can import the "main" using that bare `lodash` specifier?

Reply:
Yes, the main would in theory support a querystring itself, but not the package name, which could for example be defined as `specifier.match(/^([^\/]+)(\/|$)/)`.

Reply:
Made a tracking issue for clarification, I think this is a package name map bug personally but want feedback. https://github.com/domenic/package-name-maps/issues/38

Reply:
the champion has no desire to change the behavior, we should treat loading `fs?x` as not resolving to a builtin module probably in light of this and should throw errors.

Reply:
we might want to reserve the builtin names as prefixes though that always error to avoid some confusing behavior described in https://github.com/nodejs/TSC/issues/389